Transaction-based SQL docs

sploiselle opened the pull request:
Contains documents for:
- `BEGIN`
- `COMMIT`
- `SET TRANSACTION`
- `SAVEPOINT cockroach_restart`
- `RELEASE SAVEPOINT cockroach_restart`
- `ROLLBACK`
- `SHOW TRANSACTION`

And integrates these links into `transactions.html`

Closes #491, #254, #253

<!-- Reviewable:start -->
---

This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/docs/672)

<!-- Reviewable:end -->

sploiselle commented:
@andreimatei Per our offline discussion, could I get your help to understand when should you use `SAVEPOINT` vs. when does it not provide any benefit?

bdarnell commented on begin-transaction.md:
```diff
@@ -1,20 +1,19 @@
-The `BEGIN` [statement](sql-statements.html) initiates a [transaction](transactions.html), which bundles multiple SQL statements into a single all-or-nothing transaction. 
+The `BEGIN` [statement](sql-statements.html) initiates a [transaction](transactions.html), which either successfully executes all of the statements it contains or none at all.
+
+{{site.data.alerts.callout_info}}If the transaction contends for the same resources as other transactions and "loses", you can <a href="transactions.html#transaction-retries">retry it on the client side</a> using <a href="savepoint.html"><code>SAVEPOINT</code></a> (shown in the <a href="#examples">examples below</a>).{{site.data.alerts.end}}
```
You can (and should) retry the transaction whether you use SAVEPOINT or not; SAVEPOINT makes the retries more efficient. 

We should have a section that explains optimistic vs pessimistic concurrency; I think links like this should point to that explanation and then we link from there to SAVEPOINT.

bdarnell commented on begin-transaction.md:
```diff
@@ -24,16 +23,66 @@ No [privileges](privileges.html) are required to initiate a transaction. However
 |-----------|-------------|
-|  |  |
+| `ISOLATION LEVEL` | If you don't want the transaction to run as `SERIALIZABLE` (CockroachDB's default, which provides the highest level of isolation), you can specify `SNAPSHOT`, which can provide better performance in high-contention scenarios.<br/><br/>For more information, see [Transactions: Isolation Levels](transactions.html#isolation-levels).<br/><br/>**Default**: `SERIALIZABLE` |
+| `PRIORITY` | If you don't want the transaction to run with `NORMAL` priority, you can set it to `LOW` or `HIGH`.<br/><br/>When transactions contend for access to the same resources, the transaction with the highest priority executes first; the lower priority transaction is then retried inheriting the priority of the "winning" transaction. This means the longer a transaction contends for resourc ...
```
I'd leave out details like the inheritance of priorities here and just say that high priority transactions are less likely to be restarted than low priority ones. (but it's not guaranteed - low priority transactions can sometimes win). Save the rest of the details for the priorities section.

bdarnell commented on commit-transaction.md:
```diff
@@ -1,27 +1,55 @@
+
+> INSERT INTO orders (customer, status) VALUES (1001, 'new');
+
+> RELEASE SAVEPOINT cockroach_restart;
```
If you're using SAVEPOINT, then `RELEASE SAVEPOINT` is the real "commit"; the actual `COMMIT` statement does nothing but prepare the connection for the next transaction (things are structured this way for compatibility with other databases and client frameworks). This is surprising and needs to be spelled out clearly.

bdarnell commented on set-transaction.md:
```diff
@@ -14,12 +14,44 @@ The `SET TRANSACTION` [statement](sql-statements.html) sets the transaction isol
+> COMMIT;
+~~~
+
+{{site.data.alerts.callout_info}}If you are using <code>SAVEPOINT</code>, which we recommend, you must execute it immediately after executing <code>BEGIN</code>.<br/><br/>Your application also needs to include functions to retry transactions contending for the same resources using <a href="rollback-transaction.html#retry-a-transaction"><code>ROLLBACK TO cockroach_restart</code></a>.{{site.data.al ...
```
"Immediately after BEGIN" is incorrect if `SET TRANSACTION` statements are used.

andreimatei commented:
so per our discussion: the client doesn't need to worry about retries (because they're done automatically for):
- statements outside transactions (aka implicit transactions)
- batches of statements that contain all the transaction: `BEGIN; ... END;` (here my explanation was confused before - there should always a way to ask a driver to send a bunch of statements as one batch)
- (if you want to get even more technical) prefixes of transactions: if the server receives `BEGIN; foo; bar;`, we will automatically retry foo and bar

---

Review status: 0 of 16 files reviewed at latest revision, 9 unresolved discussions.

---

*[begin-transaction.md, line 9 at r1](https://reviewable.io:443/reviews/cockroachdb/docs/672#-KSb8fhDaAXXZT9MLQ2E:-KSb8fhDaAXXZT9MLQ2F:b-3zc0r5) ([raw file](https://github.com/cockroachdb/docs/blob/e8d155cf54f86ca5430592ca8a1784b7967643cc/begin-transaction.md#L9)):*
> ```Markdown
> The `BEGIN` [statement](sql-statements.html) initiates a [transaction](transactions.html), which either successfully executes all of the statements it contains or none at all.
> 
> {{site.data.alerts.callout_info}}If the transaction contends for the same resources as other transactions and "loses", you can <a href="transactions.html#transaction-retries">retry it on the client side</a> using <a href="savepoint.html"><code>SAVEPOINT</code></a> (shown in the <a href="#examples">examples below</a>).{{site.data.alerts.end}}
> ```

I don't know if we should use the word "contends for resources" here. Might be better to say smth like "the read/write set of the transaction overlaps with another concurrent/recent txn", so it's more suggestive.

---

*[begin-transaction.md, line 9 at r1](https://reviewable.io:443/reviews/cockroachdb/docs/672#-KSJzSXaruocoNSI46v--r1-9:-KSb9F9miBp8naQh7Nwl:blz2dx9) ([raw file](https://github.com/cockroachdb/docs/blob/e8d155cf54f86ca5430592ca8a1784b7967643cc/begin-transaction.md#L9)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary>
> You can (and should) retry the transaction whether you use SAVEPOINT or not; SAVEPOINT makes the retries more efficient. 
> 
> 
> 
> We should have a section that explains optimistic vs pessimistic concurrency; I think links like this should point to that explanation and then we link from there to SAVEPOINT. 
</details>
Indeed, using `SAVEPOINT` just allows the database to do some optimizations, and increases your odds of succeeding the next time. It changes the database behavior to not destroy all transaction state when a retryable error is thrown.

"retry it on the client side" -> "the client should retry it"

---

*[begin-transaction.md, line 62 at r1](https://reviewable.io:443/reviews/cockroachdb/docs/672#-KSbAjcVl7hL1WvSRYKx:-KSbAjcWhmg5USiXSeRw:box9dp2) ([raw file](https://github.com/cockroachdb/docs/blob/e8d155cf54f86ca5430592ca8a1784b7967643cc/begin-transaction.md#L62)):*
> ```Markdown
> ~~~
> 
> {{site.data.alerts.callout_info}}If you are using <code>SAVEPOINT</code>, which we recommend, you must execute it immediately after executing <code>BEGIN</code>.<br/><br/>Your application also needs to include functions to retry transactions contending for the same resources using <a href="rollback-transaction.html#retry-a-transaction"><code>ROLLBACK TO cockroach_restart</code></a>.{{site.data.alerts.end}}
> ```

I'd replace "you/your application" with "the client".

---

*[begin-transaction.md, line 62 at r1](https://reviewable.io:443/reviews/cockroachdb/docs/672#-KSbAxRliwdcOet2mqrV:-KSbAxRliwdcOet2mqrW:bupa424) ([raw file](https://github.com/cockroachdb/docs/blob/e8d155cf54f86ca5430592ca8a1784b7967643cc/begin-transaction.md#L62)):*
> ```Markdown
> ~~~
> 
> {{site.data.alerts.callout_info}}If you are using <code>SAVEPOINT</code>, which we recommend, you must execute it immediately after executing <code>BEGIN</code>.<br/><br/>Your application also needs to include functions to retry transactions contending for the same resources using <a href="rollback-transaction.html#retry-a-transaction"><code>ROLLBACK TO cockroach_restart</code></a>.{{site.data.alerts.end}}
> ```

maybe we should also add the link to `[transaction code samples]` here

---

*[commit-transaction.md, line 9 at r1](https://reviewable.io:443/reviews/cockroachdb/docs/672#-KSbBmwxqvwbxCdTs0qu:-KSbBmwxqvwbxCdTs0qv:bvrer2q) ([raw file](https://github.com/cockroachdb/docs/blob/e8d155cf54f86ca5430592ca8a1784b7967643cc/commit-transaction.md#L9)):*
> ```Markdown
> The `COMMIT` statement commits the current [transaction](transactions.html).
> 
> If the transaction contends for the same resources as other transactions and "loses" when it commits, you can [retry the transaction on the client side](transactions.html#transaction-retries) using [`SAVEPOINT`](savepoint.html) (shown in the [example below](#example)).
> ```

what Ben said below - if `RELEASE SAVEPOINT` is used, `COMMIT` no longer does anything. It just lets you start a new transaction on that connection.

---

*[commit-transaction.md, line 11 at r1](https://reviewable.io:443/reviews/cockroachdb/docs/672#-KSbBxlLtZol-scY06Wy:-KSbBxlLtZol-scY06Wz:b-vfxl9t) ([raw file](https://github.com/cockroachdb/docs/blob/e8d155cf54f86ca5430592ca8a1784b7967643cc/commit-transaction.md#L11)):*
> ```Markdown
> If the transaction contends for the same resources as other transactions and "loses" when it commits, you can [retry the transaction on the client side](transactions.html#transaction-retries) using [`SAVEPOINT`](savepoint.html) (shown in the [example below](#example)).
> 
> If the transaction generates any other types of errors, committing it causes a [rollback](rollback-transaction.html), which discards *all* updates made by statements in the transaction.
> ```

except when `RELEASE` was used. In that case, error on `COMMIT` no longer rolls back anything. And it also can't return retryable errors.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/docs/672#-:-KSbCdipTOISIkTnbWIh:brgv4m)*
<!-- Sent from Reviewable.io -->

jseldess commented:
@sploiselle, can you add `SHOW TRANSACTION` to this PR as well?

cockroach-teamcity commented:
http://cockroach-docs-review.s3-website-us-east-1.amazonaws.com/d0763eaf97a668ca7026d50db107c5427829d125/

sploiselle commented:
@bdarnell @andreimatei Substantially changed `transactions.md`, as well as integrated that information into the rest of the docs. Believe this also addresses the other outstanding issues. Could I please get another set of reviews on this PR?

cockroach-teamcity commented:
http://cockroach-docs-review.s3-website-us-east-1.amazonaws.com/a150b517dfd2fbf84711f8cc1bf4c872dd775adf/

bdarnell commented on transactions.md:
```diff
@@ -12,90 +12,169 @@ CockroachDB supports bundling multiple SQL statements into a single all-or-nothi
+
+You can find examples of this in the [Syntax](#syntax) section of this page or at our page [Build a Test App: Execute transaction from a client](build-a-test-app.html#step-4-execute-transactions-from-a-client).
+
+{{site.data.alerts.callout_success}}For Golang users, client-side retries are built into the <code>crdb</code> package of the CockroachDB Go client and happen automatically. For more information, see the <strong>Go</strong> tab of <a href="build-a-test-app.html#step-4-execute-transactions-from-a-client">Build a Test App: Execute transaction from a client</a>.{{site.data.alerts.end}}
```
Go, not Golang. We also have this functionality for the SQLAlchemy package in python.

cockroach-teamcity commented:
http://cockroach-docs-review.s3-website-us-east-1.amazonaws.com/b8eee8c1008200317fc543368bd735fc2bcfbcf1/

andreimatei commented on transactions.md:
```diff
@@ -12,90 +12,173 @@ CockroachDB supports bundling multiple SQL statements into a single all-or-nothi
+
+Implementing client-side retries requires three statements:
+
+- [`SAVEPOINT cockroach_restart`](savepoint.html) defines your intent to retry the transaction if there are contention errors. It must be executed after `BEGIN` but before the first statement that manipulates a database.
```
s/defines/declares

andreimatei replied:
s/your/the client's ?

andreimatei commented on transactions.md:
```diff
@@ -12,90 +12,173 @@ CockroachDB supports bundling multiple SQL statements into a single all-or-nothi
+Implementing client-side retries requires three statements:
+
+- [`SAVEPOINT cockroach_restart`](savepoint.html) defines your intent to retry the transaction if there are contention errors. It must be executed after `BEGIN` but before the first statement that manipulates a database.
+- [`ROLLBACK TO cockroach_restart`](rollback-transaction.html#retry-a-transaction) is used when your application detects `40001` / `retry transaction` errors. It provides you a chance to "retry" the transaction by rolling the database's state back to the beginning of the transaction and incrementing the transaction's priority. 
```
s/incrementing/ratcheting up

andreimatei commented on transactions.md:
```diff
@@ -12,90 +12,173 @@ CockroachDB supports bundling multiple SQL statements into a single all-or-nothi
+- [`SAVEPOINT cockroach_restart`](savepoint.html) defines your intent to retry the transaction if there are contention errors. It must be executed after `BEGIN` but before the first statement that manipulates a database.
+- [`ROLLBACK TO cockroach_restart`](rollback-transaction.html#retry-a-transaction) is used when your application detects `40001` / `retry transaction` errors. It provides you a chance to "retry" the transaction by rolling the database's state back to the beginning of the transaction and incrementing the transaction's priority. 
+  
+  After issuing `ROLLBACK TO cockroach_restart`, each of the transaction's statements must then be executed again.
```
"must be executed again" - well, not technically. The client can issue whatever statements they want. But we are optimized for the case when they're gonna attempt mostly the same writes. This is an important, but subtle, point - the conditional logic one. The client is supposed to execute what is morally the same transaction as before. But they're free to alter it in any way, shape, or form (including issuing completely different statements to the point where the new iteration of the txn appears to the database to be completely unrelated to what had been attempted before).

andreimatei commented on transactions.md:
```diff
@@ -12,90 +12,173 @@ CockroachDB supports bundling multiple SQL statements into a single all-or-nothi
+  
+  After issuing `ROLLBACK TO cockroach_restart`, each of the transaction's statements must then be executed again.
+
+- [`RELEASE SAVEPOINT cockroach_restart`](release-savepoint.html) commits the transaction.
```
a think a note here explaining why this RELEASE is necessary would be helpful: it gives the client the chance to trap a "retryable" error that can happen when attempting to commit. And, in fact, when attempting to commit is one of the more likely places when the client might find out that it needs to retry, because of implementation details: sometimes the transaction finds out late that it has been "pushed" by another one. And sometimes we even let transactions that we know will have to be eventually restarted run for a while, so they lay out more intents and thus have a higher chance of succeeding the next time around.

And the thing is `COMMIT` does not allow a client to receive this errors and act on them. For one, it's often not under the control of the client, as it's issued by a driver. And second, there's no recourse from COMMIT, at least not with the postgres protocol.

andreimatei commented on transactions.md:
```diff
@@ -12,90 +12,173 @@ CockroachDB supports bundling multiple SQL statements into a single all-or-nothi
+
+1. The transaction starts with the `BEGIN` statement.
+
+2. The `SAVEPOINT cockroach_restart` statement defines the intention to retry the transaction in the case of contention errors. Note that CockroachDB's savepoint implementation does not support all savepoint functionality, such as nested transactions. 
```
s/define/declare

andreimatei commented on transactions.md:
```diff
@@ -12,90 +12,173 @@ CockroachDB supports bundling multiple SQL statements into a single all-or-nothi
+CockroachDB uses slightly different isolation levels than [ANSI SQL isolation levels](https://en.wikipedia.org/wiki/Isolation_(database_systems)#Isolation_levels).
 
-The CockroachDB `SNAPSHOT ISOLATION` level is stronger than the ANSI SQL `READ UNCOMMITTED` and `READ COMMITTED` levels.
+- The CockroachDB `SERIALIZABLE` level is stronger than the ANSI SQL `REPEATABLE READ` level and equivalent to the ANSI SQL `SERIALIZABLE` level.
```
I think we used to have a note once about how REPEATABLE READ is neither stronger nor weaker nor the same as SNAPSHOT. I don't know why it was removed, but I think it was good to have.

andreimatei commented:
A lot of work here, looks good! Random comments for you to pick and choose.

sploiselle commented:
Review status: 0 of 17 files reviewed at latest revision, 28 unresolved discussions.

---

*[begin-transaction.md, line 9 at r1](https://reviewable.io:443/reviews/cockroachdb/docs/672#-KSJzSXaruocoNSI46v--r1-9:-KUSOz1ePmIMHLUF73-q:b-896fix) ([raw file](https://github.com/cockroachdb/docs/blob/e8d155cf54f86ca5430592ca8a1784b7967643cc/begin-transaction.md#L9)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary>
> Indeed, using `SAVEPOINT` just allows the database to do some optimizations, and increases your odds of succeeding the next time. It changes the database behavior to not destroy all transaction state when a retryable error is thrown.
> 
> "retry it on the client side" -> "the client should retry it"
</details>
Done.

---

*[begin-transaction.md, line 9 at r1](https://reviewable.io:443/reviews/cockroachdb/docs/672#-KSb8fhDaAXXZT9MLQ2E:-KUSP28ZEmHylSb9blIu:b-896fix) ([raw file](https://github.com/cockroachdb/docs/blob/e8d155cf54f86ca5430592ca8a1784b7967643cc/begin-transaction.md#L9)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary>
> I don't know if we should use the word "contends for resources" here. Might be better to say smth like "the read/write set of the transaction overlaps with another concurrent/recent txn", so it's more suggestive.
</details>
Done.

---

*[begin-transaction.md, line 40 at r1](https://reviewable.io:443/reviews/cockroachdb/docs/672#-KSJzSXaruocoNSI46v--r1-40:-KUSPOJz9Dn_YvIdJvLa:b-896fix) ([raw file](https://github.com/cockroachdb/docs/blob/e8d155cf54f86ca5430592ca8a1784b7967643cc/begin-transaction.md#L40)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary>
> I'd leave out details like the inheritance of priorities here and just say that high priority transactions are less likely to be restarted than low priority ones. (but it's not guaranteed - low priority transactions can sometimes win). Save the rest of the details for the priorities section. 
</details>
Done.

---

*[begin-transaction.md, line 62 at r1](https://reviewable.io:443/reviews/cockroachdb/docs/672#-KSbAjcVl7hL1WvSRYKx:-KUSQJ-omGCIuArkPEr8:b-896fix) ([raw file](https://github.com/cockroachdb/docs/blob/e8d155cf54f86ca5430592ca8a1784b7967643cc/begin-transaction.md#L62)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary>
> I'd replace "you/your application" with "the client".
</details>
Done.

---

*[begin-transaction.md, line 62 at r1](https://reviewable.io:443/reviews/cockroachdb/docs/672#-KSbAxRliwdcOet2mqrV:-KUSQM96hrc1J_Mi_5tY:b-896fix) ([raw file](https://github.com/cockroachdb/docs/blob/e8d155cf54f86ca5430592ca8a1784b7967643cc/begin-transaction.md#L62)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary>
> maybe we should also add the link to `[transaction code samples]` here
</details>
Done.

---

*[begin-transaction.md, line 9 at r3](https://reviewable.io:443/reviews/cockroachdb/docs/672#-KSJzSXaruocoNSI46v--r3-9:-KUSOyHZgT3St0JEAMtc:b-896fix) ([raw file](https://github.com/cockroachdb/docs/blob/b8eee8c1008200317fc543368bd735fc2bcfbcf1/begin-transaction.md#L9)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary>
> s/must/should ?
</details>
Done.

---

*[begin-transaction.md, line 30 at r3](https://reviewable.io:443/reviews/cockroachdb/docs/672#-KSJzSXaruocoNSI46v--r3-30:-KUSPKr-jEQD5BJ__UBq:b-mnv0zn) ([raw file](https://github.com/cockroachdb/docs/blob/b8eee8c1008200317fc543368bd735fc2bcfbcf1/begin-transaction.md#L30)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary>
> isolation levels haven't been introduced yet on this page. You sure you want to put this here?
</details>
Added link to isolation levels. Great catch.

---

*[begin-transaction.md, line 57 at r3](https://reviewable.io:443/reviews/cockroachdb/docs/672#-KSJzSXaruocoNSI46v--r3-57:-KUSQGWUmURtitIPGEVI:b-896fix) ([raw file](https://github.com/cockroachdb/docs/blob/b8eee8c1008200317fc543368bd735fc2bcfbcf1/begin-transaction.md#L57)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary>
> this example doesn't seem very suggestive to me. Maybe make the insert:
> 
> `INSERT INTO orders (customer, sku, status) VALUES (101, '867...', 'new')
> 
> Feel free to ignore if we use this schema elsewhere.
</details>
Done.

---

*[begin-transaction.md, line 71 at r3](https://reviewable.io:443/reviews/cockroachdb/docs/672#-KSJzSXaruocoNSI46v--r3-71:-KUSQkpmNGU_mnTOJpMw:b-896fix) ([raw file](https://github.com/cockroachdb/docs/blob/b8eee8c1008200317fc543368bd735fc2bcfbcf1/begin-transaction.md#L71)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary>
> you can also set these with a `SET` stmt, and I think it's worth pointing that out since sometimes a user doesn't have access to controlling the `BEGIN` stmt as that's done by a driver.
> 
> 
> 
> Ah, now I see that has a special page. I think it might be worth linking from here.
</details>
Done.

---

*[commit-transaction.md, line 9 at r1](https://reviewable.io:443/reviews/cockroachdb/docs/672#-KSbBmwxqvwbxCdTs0qu:-KUSR8Y6prfKfHroPjdn:b-896fix) ([raw file](https://github.com/cockroachdb/docs/blob/e8d155cf54f86ca5430592ca8a1784b7967643cc/commit-transaction.md#L9)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary>
> what Ben said below - if `RELEASE SAVEPOINT` is used, `COMMIT` no longer does anything. It just lets you start a new transaction on that connection.
</details>
Done.

---

*[commit-transaction.md, line 11 at r1](https://reviewable.io:443/reviews/cockroachdb/docs/672#-KSbBxlLtZol-scY06Wy:-KUOTRuLUOpbi271L-F_:b-896fix) ([raw file](https://github.com/cockroachdb/docs/blob/e8d155cf54f86ca5430592ca8a1784b7967643cc/commit-transaction.md#L11)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary>
> except when `RELEASE` was used. In that case, error on `COMMIT` no longer rolls back anything. And it also can't return retryable errors.
</details>
Done.

---

*[commit-transaction.md, line 42 at r1](https://reviewable.io:443/reviews/cockroachdb/docs/672#-KSJzSXaruocoNSI46v0-r1-42:-KUOTQYbJrBa_NwzooX-:b-896fix) ([raw file](https://github.com/cockroachdb/docs/blob/e8d155cf54f86ca5430592ca8a1784b7967643cc/commit-transaction.md#L42)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary>
> If you're using SAVEPOINT, then `RELEASE SAVEPOINT` is the real "commit"; the actual `COMMIT` statement does nothing but prepare the connection for the next transaction (things are structured this way for compatibility with other databases and client frameworks). This is surprising and needs to be spelled out clearly.
</details>
Done.

---

*[commit-transaction.md, line 11 at r3](https://reviewable.io:443/reviews/cockroachdb/docs/672#-KSJzSXaruocoNSI46v0-r3-11:-KUSRO0qi_vWwnd5YL1Q:b-896fix) ([raw file](https://github.com/cockroachdb/docs/blob/b8eee8c1008200317fc543368bd735fc2bcfbcf1/commit-transaction.md#L11)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary>
> "committing causes a rollback" -> "`COMMIT` is equivalent to `ROLLBACK`"
</details>
Done.

---

*[release-savepoint.md, line 2 at r3](https://reviewable.io:443/reviews/cockroachdb/docs/672#-KSJzSXaruocoNSI46v2-r3-2:-KUSS6wBbkPAaod_uL9d:b-896fix) ([raw file](https://github.com/cockroachdb/docs/blob/b8eee8c1008200317fc543368bd735fc2bcfbcf1/release-savepoint.md#L2)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary>
> should `cockroach_restart` be in the title?
> 
> 
> 
> And on a related note, I think it'd be nice if this page had the same "danger" bubble that another one had about how Cockroach doesn't exactly have general support for savepoints.
</details>
Done.

---

*[set-transaction.md, line 48 at r1](https://reviewable.io:443/reviews/cockroachdb/docs/672#-KSJzSXaruocoNSI46v5-r1-48:-KUSS8qUojB53fgVGUr1:b-896fix) ([raw file](https://github.com/cockroachdb/docs/blob/e8d155cf54f86ca5430592ca8a1784b7967643cc/set-transaction.md#L48)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary>
> "Immediately after BEGIN" is incorrect if `SET TRANSACTION` statements are used. 
</details>
Done.

---

*[transactions.md, line 110 at r2](https://reviewable.io:443/reviews/cockroachdb/docs/672#-KSJzSXaruocoNSI46v6-r2-110:-KUSX56Nt_FU-LeTbf2g:b-896fix) ([raw file](https://github.com/cockroachdb/docs/blob/a150b517dfd2fbf84711f8cc1bf4c872dd775adf/transactions.md#L110)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary>
> Go, not Golang. We also have this functionality for the SQLAlchemy package in python. 
</details>
Done.

---

*[transactions.md, line 71 at r3](https://reviewable.io:443/reviews/cockroachdb/docs/672#-KSJzSXaruocoNSI46v6-r3-71:-KUSgq0FkSgfPbfkNPSs:b-896fix) ([raw file](https://github.com/cockroachdb/docs/blob/b8eee8c1008200317fc543368bd735fc2bcfbcf1/transactions.md#L71)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary>
> ditto "separate lines" and s/must/should
</details>
Done.

---

*[transactions.md, line 87 at r3](https://reviewable.io:443/reviews/cockroachdb/docs/672#-KSJzSXaruocoNSI46v6-r3-87:-KUSclP8a5U21v8FbPvz:b-896fix) ([raw file](https://github.com/cockroachdb/docs/blob/b8eee8c1008200317fc543368bd735fc2bcfbcf1/transactions.md#L87)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary>
> what are `cockroach_restart` functions? Maybe say `SAVEPOINT cockroach_restart`.
</details>
Done.

---

*[transactions.md, line 99 at r3](https://reviewable.io:443/reviews/cockroachdb/docs/672#-KSJzSXaruocoNSI46v6-r3-99:-KUSckYy703oNQhWc_m5:b-896fix) ([raw file](https://github.com/cockroachdb/docs/blob/b8eee8c1008200317fc543368bd735fc2bcfbcf1/transactions.md#L99)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary>
> s/your/the client's ?
</details>
Done.

---

*[transactions.md, line 100 at r3](https://reviewable.io:443/reviews/cockroachdb/docs/672#-KSJzSXaruocoNSI46v6-r3-100:-KUSciyztaRaeg2ZwXZS:b-896fix) ([raw file](https://github.com/cockroachdb/docs/blob/b8eee8c1008200317fc543368bd735fc2bcfbcf1/transactions.md#L100)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary>
> s/incrementing/ratcheting up
</details>
Done.

---

*[transactions.md, line 102 at r3](https://reviewable.io:443/reviews/cockroachdb/docs/672#-KSJzSXaruocoNSI46v6-r3-102:-KUSXwMNFY7vOOaoWcil:b5ilov0) ([raw file](https://github.com/cockroachdb/docs/blob/b8eee8c1008200317fc543368bd735fc2bcfbcf1/transactions.md#L102)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary>
> "must be executed again" - well, not technically. The client can issue whatever statements they want. But we are optimized for the case when they're gonna attempt mostly the same writes. This is an important, but subtle, point - the conditional logic one. The client is supposed to execute what is morally the same transaction as before. But they're free to alter it in any way, shape, or form (including issuing completely different statements to the point where the new iteration of the txn appears to the database to be completely unrelated to what had been attempted before).
</details>
Clarified this.

---

*[transactions.md, line 104 at r3](https://reviewable.io:443/reviews/cockroachdb/docs/672#-KSJzSXaruocoNSI46v6-r3-104:-KUSWaknaEqoUIu-YMam:bb615du) ([raw file](https://github.com/cockroachdb/docs/blob/b8eee8c1008200317fc543368bd735fc2bcfbcf1/transactions.md#L104)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary>
> a think a note here explaining why this RELEASE is necessary would be helpful: it gives the client the chance to trap a "retryable" error that can happen when attempting to commit. And, in fact, when attempting to commit is one of the more likely places when the client might find out that it needs to retry, because of implementation details: sometimes the transaction finds out late that it has been "pushed" by another one. And sometimes we even let transactions that we know will have to be eventually restarted run for a while, so they lay out more intents and thus have a higher chance of succeeding the next time around.
> 
> 
> 
> And the thing is `COMMIT` does not allow a client to receive this errors and act on them. For one, it's often not under the control of the client, as it's issued by a driver. And second, there's no recourse from COMMIT, at least not with the postgres protocol.
</details>
I think this kind of detail would make an amazing blog post. It's a little hard to work into the documentation, though. I added a line mentioning that executing `RELEASE SAVEPOINT` performs the contention checks.

---

*[transactions.md, line 120 at r3](https://reviewable.io:443/reviews/cockroachdb/docs/672#-KSJzSXaruocoNSI46v6-r3-120:-KUSX8Rhr9g9Ol4s_BRK:b-896fix) ([raw file](https://github.com/cockroachdb/docs/blob/b8eee8c1008200317fc543368bd735fc2bcfbcf1/transactions.md#L120)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary>
> s/define/declare
</details>
Done.

---

*[transactions.md, line 180 at r3](https://reviewable.io:443/reviews/cockroachdb/docs/672#-KSJzSXaruocoNSI46v6-r3-180:-KUSXR72s-yGozm-RMKp:b488q7z) ([raw file](https://github.com/cockroachdb/docs/blob/b8eee8c1008200317fc543368bd735fc2bcfbcf1/transactions.md#L180)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary>
> I think we used to have a note once about how REPEATABLE READ is neither stronger nor weaker nor the same as SNAPSHOT. I don't know why it was removed, but I think it was good to have.
</details>
That it's not stronger, weaker, or the same is hard to contextualize. Could I get a brief description of their differences?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/docs/672)*
<!-- Sent from Reviewable.io -->

sploiselle commented:
@jseldess Ready for your review.

cockroach-teamcity commented:
http://cockroach-docs-review.s3-website-us-east-1.amazonaws.com/e92f413328c1292ef88d48cbb8bc49c759916264/

jseldess commented:
Review status: 0 of 17 files reviewed at latest revision, 39 unresolved discussions, some commit checks failed.

---

*[begin-transaction.md, line 62 at r1](https://reviewable.io:443/reviews/cockroachdb/docs/672#-KSbAjcVl7hL1WvSRYKx:-KUrnUnt4G75AHHH5Kpk:b7xr63c) ([raw file](https://github.com/cockroachdb/docs/blob/e8d155cf54f86ca5430592ca8a1784b7967643cc/begin-transaction.md#L62)):*
<details><summary><i>Previously, sploiselle (Sean Loiselle) wrote…</i></summary>
> Done.
</details>
Would it be possible to include a concise explanation of why in this case you need client-side retry? Maybe something like: "Because the statements in this transaction are not all on a single line, this transaction would require ...."

---

*[begin-transaction.md, line 86 at r4](https://reviewable.io:443/reviews/cockroachdb/docs/672#-KUrnzPPSMLaiGUUo_Mf:-KUrnzPPSMLaiGUUo_Mg:by27n98) ([raw file](https://github.com/cockroachdb/docs/blob/e92f413328c1292ef88d48cbb8bc49c759916264/begin-transaction.md#L86)):*
> ```Markdown
> You can also set a transaction's isolation level and priority with [`SET TRANSACTION`](set-transaction.html).
> 
> {{site.data.alerts.callout_danger}}This examples requires <a href="transactions.html#client-side-intervention">client-side intervention to handle transaction retries</a>.{{site.data.alerts.end}}
> ```

See comment above about explaining why client-side retry is required.

---

*[commit-transaction.md, line 11 at r4](https://reviewable.io:443/reviews/cockroachdb/docs/672#-KUrs2MIJJXw4CF-lzje:-KUrs2MIJJXw4CF-lzjf:b-aypkvd) ([raw file](https://github.com/cockroachdb/docs/blob/e92f413328c1292ef88d48cbb8bc49c759916264/commit-transaction.md#L11)):*
> ```Markdown
> When using [client-side transaction retries](transactions.html#client-side-transaction-retries), statements issued after [`SAVEPOINT cockroach_restart`](savepoint.html) are committed when [`RELEASE SAVEPOINT cockroach_restart`](release-savepoint.html) is issued instead of `COMMIT`. However, you must still issue a `COMMIT` statement to clear the connection for the next transaction.
> 
> If statements in the transaction generated any errors, `COMMIT` is equivalent to `ROLLBACK`, which aborts the transaction and discards *all* updates made by its statements.
> ```

As discussed offline, I don't think it's clear that this paragraph applies only to the non-retryable scenario.

---

*[commit-transaction.md, line 51 at r4](https://reviewable.io:443/reviews/cockroachdb/docs/672#-KUrqMfdh4wFVGK-2wqu:-KUrqMfeVSKgCHLRSXbv:bb2gqqs) ([raw file](https://github.com/cockroachdb/docs/blob/e92f413328c1292ef88d48cbb8bc49c759916264/commit-transaction.md#L51)):*
> ```Markdown
> ~~~
> 
> {{site.data.alerts.callout_danger}}This examples requires <a href="transactions.html#client-side-intervention">client-side intervention to handle transaction retries</a>.{{site.data.alerts.end}}
> ```

Same feedback as on `BEGIN`: It might be helpful to include a concise explanation of why in this case you need client-side retry, e.g., "Because the statements in this transaction are not all on a single line, this transaction would require ...."

---

*[release-savepoint.md, line 7 at r4](https://reviewable.io:443/reviews/cockroachdb/docs/672#-KUrsHxaX9Um_d2Yf57V:-KUrsHxaX9Um_d2Yf57W:b-weufv9) ([raw file](https://github.com/cockroachdb/docs/blob/e92f413328c1292ef88d48cbb8bc49c759916264/release-savepoint.md#L7)):*
> ```Markdown
> ---
> 
> When using [client-side transaction retries](transactions.html#client-side-transaction-retries), `RELEASE SAVEPOINT cockroach_restart` statement commits the transaction.
> ```

Add article: "The `RELEASE SAVEPOINT cockroach_restart` statement..."

---

*[release-savepoint.md, line 27 at r4](https://reviewable.io:443/reviews/cockroachdb/docs/672#-KUrsXJonsyxNYxovt2e:-KUrsXJpAwIyMwC-Nx_j:b6qh4ns) ([raw file](https://github.com/cockroachdb/docs/blob/e92f413328c1292ef88d48cbb8bc49c759916264/release-savepoint.md#L27)):*
> ```Markdown
> ### Commit a Transaction
> 
> After declaring `SAVEPOINT cockroach_restart`, commit the transactions with `RELEASE SAVEPOINT cockroach_restart` and then prepare the connection for the next transaction with `COMMIT`.
> ```

Should "transactions" be singular?

---

*[set-transaction.md, line 3 at r4](https://reviewable.io:443/reviews/cockroachdb/docs/672#-KUrxCoxqZfg-QVkGydB:-KUrwAGTlNkF3r6LhqtR:bdqaa9b) ([raw file](https://github.com/cockroachdb/docs/blob/e92f413328c1292ef88d48cbb8bc49c759916264/set-transaction.md#L3)):*
> ```Markdown
> ---
> title: SET TRANSACTION
> summary: The SET TRANSACTION statement sets the transaction isolation level and/or priority for the current session or for an individual transaction.
> ```

As discussed, this `SET` statement isn't for the session; it's only for the current transaction, as you explain in the intro. Best to update this summary as well.

That said, we should check if postgres supports setting the isolation level/priority for an entire session. If so, we should probably do so as well. Sean, can you please look into this?

---

*[set-transaction.md, line 48 at r4](https://reviewable.io:443/reviews/cockroachdb/docs/672#-KUrvsQABWU4Hoqg1UVM:-KUrvsQABWU4Hoqg1UVN:b6p012m) ([raw file](https://github.com/cockroachdb/docs/blob/e92f413328c1292ef88d48cbb8bc49c759916264/set-transaction.md#L48)):*
> ```Markdown
> ~~~
> 
> {{site.data.alerts.callout_danger}}This examples requires <a href="transactions.html#client-side-intervention">client-side intervention to handle transaction retries</a>.{{site.data.alerts.end}}
> ```

Same comment as similar notes on other pages: let's add a brief explanation of why you need to do client-side retry.

---

*[show-transaction.md, line 8 at r4](https://reviewable.io:443/reviews/cockroachdb/docs/672#-KUrzkyp8yvWZ4qYzh84:-KUrzkyp8yvWZ4qYzh85:b-eu0bwl) ([raw file](https://github.com/cockroachdb/docs/blob/e92f413328c1292ef88d48cbb8bc49c759916264/show-transaction.md#L8)):*
> ```Markdown
> ---
> 
> The `SHOW TRANSACTION` [statement](sql-statements.html) lists the default transaction isolation level or transaction priority for the current session or for an individual [transaction](transactions.html).
> ```

Since you can't change the session default, the context determines what this statement returns: If issued outside of a transaction, it simply shows you the default. If issued within a statement, it shows you the iso level for that statement. Maybe clarify a bit?

---

*[transactions.md, line 120 at r4](https://reviewable.io:443/reviews/cockroachdb/docs/672#-KUs0JJ9j5EMrJERU4t_:-KUs0JJAo7QHTn2OF4z5:b-1t95tq) ([raw file](https://github.com/cockroachdb/docs/blob/e92f413328c1292ef88d48cbb8bc49c759916264/transactions.md#L120)):*
> ```Markdown
> To handle these types of errors you have two options:
> 
> - *Recommended*: Use the `SAVEPOINT cockroach_restart` functions to create retryable transactions. Retryable transaction can improve performance because their priority's increased each time they are retried, making them more likely to succeed the longer they're in your system. 
> ```

At start of second sentence, "Retryable transaction" should be plural. 

---

*[transactions.md, line 128 at r4](https://reviewable.io:443/reviews/cockroachdb/docs/672#-KUs0aYVqGXl5IZ-LV4C:-KUs0aYVqGXl5IZ-LV4D:bnagrn5) ([raw file](https://github.com/cockroachdb/docs/blob/e92f413328c1292ef88d48cbb8bc49c759916264/transactions.md#L128)):*
> ```Markdown
> #### Client-Side Transaction Retries
> 
> To improve the performance of transactions, CockroachDB include a set of statements that let you retry transaction that fail due to contention errors. Retrying transactions has the benefit of increasing their priority each time they're retried, increasing their likelihood to succeed.
> ```

In first sentence:
- change "CockroachDB include" to "CockroachDB include**s**"
- change "retry transaction" to "retry transaction**s**"

---

*[transactions.md, line 218 at r4](https://reviewable.io:443/reviews/cockroachdb/docs/672#-KUs1p_zfqkhAH5n2CJ6:-KUs1p_zfqkhAH5n2CJ7:bunfvp5) ([raw file](https://github.com/cockroachdb/docs/blob/e92f413328c1292ef88d48cbb8bc49c759916264/transactions.md#L218)):*
> ```Markdown
> - The CockroachDB `SNAPSHOT` level is stronger than the ANSI SQL `READ UNCOMMITTED` and `READ COMMITTED` levels. 
> 
> {{site.data.alerts.callout_success}}<code>REPEATABLE READ</code> does <em>not</em> equate to <code>SNAPSHOT</code> in CockroachDB. We made this choice to avoid potential confusion between them and the anomalies they can introduce. <code>REPEATABLE READ</code> permits the <a href="https://en.wikipedia.org/wiki/Isolation_(database_systems)#Phantom_reads">phantom read</a> anomaly, while <code>SNAPSHOT</code> permits the <a href="https://en.wikipedia.org/wiki/Snapshot_isolation">write skew</a> anomaly.{{site.data.alerts.end}}
> ```

The confusing thing here, to me, is that `REPEATABLE READ` is a synonym for `SNAPSHOT`. I think we nee d to start this section with a clear statement about aliases, as you've done in the `BEGIN` docs, and then go on to explain that, though aliases, they are no equivalent. Thoughts?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/docs/672)*
<!-- Sent from Reviewable.io -->

sploiselle commented:
Review status: 0 of 17 files reviewed at latest revision, 39 unresolved discussions.

---

*[begin-transaction.md, line 62 at r1](https://reviewable.io:443/reviews/cockroachdb/docs/672#-KSbAjcVl7hL1WvSRYKx:-KUsErxcTEmZU_HAtGtk:b-896fix) ([raw file](https://github.com/cockroachdb/docs/blob/e8d155cf54f86ca5430592ca8a1784b7967643cc/begin-transaction.md#L62)):*
<details><summary><i>Previously, jseldess wrote…</i></summary>
> Would it be possible to include a concise explanation of why in this case you need client-side retry? Maybe something like: "Because the statements in this transaction are not all on a single line, this transaction would require ...."
</details>
Done.

---

*[begin-transaction.md, line 86 at r4](https://reviewable.io:443/reviews/cockroachdb/docs/672#-KUrnzPPSMLaiGUUo_Mf:-KUsEsi61NxqL78Mky9u:b-896fix) ([raw file](https://github.com/cockroachdb/docs/blob/e92f413328c1292ef88d48cbb8bc49c759916264/begin-transaction.md#L86)):*
<details><summary><i>Previously, jseldess wrote…</i></summary>
> See comment above about explaining why client-side retry is required.
</details>
Done.

---

*[commit-transaction.md, line 7 at r3](https://reviewable.io:443/reviews/cockroachdb/docs/672#-KSJzSXaruocoNSI46v0-r3-7:-KUs_c__n_BqeNUPja_n:b-896fix) ([raw file](https://github.com/cockroachdb/docs/blob/b8eee8c1008200317fc543368bd735fc2bcfbcf1/commit-transaction.md#L7)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary>
> one more thing that I think deserves to be spelled out here is the meaning of errors that one might get from a `COMMIT`. In particular the fact that, *if there's no RELEASE*, any kind of network error (which I guess right now you can only approximate with "any kind of internal error") should be treated as ambiguous - the transaction/data might have committed or it might not have. No way to tell directly, the client needs to read and see. This is not unique to Cockroach.
> 
> And same for errors on `RELEASE`.
</details>
Done.

---

*[commit-transaction.md, line 55 at r3](https://reviewable.io:443/reviews/cockroachdb/docs/672#-KSJzSXaruocoNSI46v0-r3-55:-KUsaHa7SS9QopFWrO1-:b-896fix) ([raw file](https://github.com/cockroachdb/docs/blob/b8eee8c1008200317fc543368bd735fc2bcfbcf1/commit-transaction.md#L55)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary>
> "you can simply COMMIT them" - as opposed to what? I don't think this phrase is very clear.
</details>
Done.

---

*[commit-transaction.md, line 11 at r4](https://reviewable.io:443/reviews/cockroachdb/docs/672#-KUrs2MIJJXw4CF-lzje:-KUsI2wZzkltckxQyHwq:b-896fix) ([raw file](https://github.com/cockroachdb/docs/blob/e92f413328c1292ef88d48cbb8bc49c759916264/commit-transaction.md#L11)):*
<details><summary><i>Previously, jseldess wrote…</i></summary>
> As discussed offline, I don't think it's clear that this paragraph applies only to the non-retryable scenario.
</details>
Done.

---

*[commit-transaction.md, line 51 at r4](https://reviewable.io:443/reviews/cockroachdb/docs/672#-KUrqMfdh4wFVGK-2wqu:-KUsaGwwjOGnGaD9ptNb:b-896fix) ([raw file](https://github.com/cockroachdb/docs/blob/e92f413328c1292ef88d48cbb8bc49c759916264/commit-transaction.md#L51)):*
<details><summary><i>Previously, jseldess wrote…</i></summary>
> Same feedback as on `BEGIN`: It might be helpful to include a concise explanation of why in this case you need client-side retry, e.g., "Because the statements in this transaction are not all on a single line, this transaction would require ...."
</details>
Done.

---

*[release-savepoint.md, line 7 at r4](https://reviewable.io:443/reviews/cockroachdb/docs/672#-KUrsHxaX9Um_d2Yf57V:-KUsaVR1-P23OaNFi3P_:b-896fix) ([raw file](https://github.com/cockroachdb/docs/blob/e92f413328c1292ef88d48cbb8bc49c759916264/release-savepoint.md#L7)):*
<details><summary><i>Previously, jseldess wrote…</i></summary>
> Add article: "The `RELEASE SAVEPOINT cockroach_restart` statement..."
</details>
Done.

---

*[release-savepoint.md, line 27 at r4](https://reviewable.io:443/reviews/cockroachdb/docs/672#-KUrsXJonsyxNYxovt2e:-KUsaVn0a62SYmh5xOvV:b-896fix) ([raw file](https://github.com/cockroachdb/docs/blob/e92f413328c1292ef88d48cbb8bc49c759916264/release-savepoint.md#L27)):*
<details><summary><i>Previously, jseldess wrote…</i></summary>
> Should "transactions" be singular?
</details>
Done.

---

*[set-transaction.md, line 3 at r4](https://reviewable.io:443/reviews/cockroachdb/docs/672#-KUrxCoxqZfg-QVkGydB:-KUsi7XOgT3KYVhZlm-Q:b-896fix) ([raw file](https://github.com/cockroachdb/docs/blob/e92f413328c1292ef88d48cbb8bc49c759916264/set-transaction.md#L3)):*
<details><summary><i>Previously, jseldess wrote…</i></summary>
> As discussed, this `SET` statement isn't for the session; it's only for the current transaction, as you explain in the intro. Best to update this summary as well.
> 
> That said, we should check if postgres supports setting the isolation level/priority for an entire session. If so, we should probably do so as well. Sean, can you please look into this?
> 
> 
</details>
Done.

---

*[set-transaction.md, line 48 at r4](https://reviewable.io:443/reviews/cockroachdb/docs/672#-KUrvsQABWU4Hoqg1UVM:-KUsiJ3lsaF3SSwVOHgv:b-896fix) ([raw file](https://github.com/cockroachdb/docs/blob/e92f413328c1292ef88d48cbb8bc49c759916264/set-transaction.md#L48)):*
<details><summary><i>Previously, jseldess wrote…</i></summary>
> Same comment as similar notes on other pages: let's add a brief explanation of why you need to do client-side retry.
</details>
Done.

---

*[show-transaction.md, line 8 at r4](https://reviewable.io:443/reviews/cockroachdb/docs/672#-KUrzkyp8yvWZ4qYzh84:-KUsjHDDwxyOsoeahX7i:b-896fix) ([raw file](https://github.com/cockroachdb/docs/blob/e92f413328c1292ef88d48cbb8bc49c759916264/show-transaction.md#L8)):*
<details><summary><i>Previously, jseldess wrote…</i></summary>
> Since you can't change the session default, the context determines what this statement returns: If issued outside of a transaction, it simply shows you the default. If issued within a statement, it shows you the iso level for that statement. Maybe clarify a bit?
</details>
Done.

---

*[transactions.md, line 61 at r3](https://reviewable.io:443/reviews/cockroachdb/docs/672#-KSJzSXaruocoNSI46v6-r3-61:-KUsjI2DTORlw9NIyPQG:b-896fix) ([raw file](https://github.com/cockroachdb/docs/blob/b8eee8c1008200317fc543368bd735fc2bcfbcf1/transactions.md#L61)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary>
> single line - one batch
</details>
Done.

---

*[transactions.md, line 67 at r3](https://reviewable.io:443/reviews/cockroachdb/docs/672#-KSJzSXaruocoNSI46v6-r3-67:-KUsjIpVM4GGQPeA5vn4:b-896fix) ([raw file](https://github.com/cockroachdb/docs/blob/b8eee8c1008200317fc543368bd735fc2bcfbcf1/transactions.md#L67)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary>
> I'm not sure about this phrase. "using the existing values" is not very clear to me... Existing where? Maybe say that the batch of statements is retried literally for as long as it takes.
> 
> 
> 
> And then the warning about "conditional values"... So the point is that we know that there's nothing conditional about these statements. Or, in other words, if there was something conditional on state that is in flux, a client already has a bug since the state, that was obviously inspected in some other txn, might have already be changed by the time this txn runs. So maybe expand this paragraph to explain this rather than leaving the current warning? Maybe phrase like "What allows CDB to automatically retry ... is the fact that all the statements and their values were sent for unconditional execution".
</details>
Done.

---

*[transactions.md, line 120 at r4](https://reviewable.io:443/reviews/cockroachdb/docs/672#-KUs0JJ9j5EMrJERU4t_:-KUsjVzk9CCr1Jy_1kAr:b-896fix) ([raw file](https://github.com/cockroachdb/docs/blob/e92f413328c1292ef88d48cbb8bc49c759916264/transactions.md#L120)):*
<details><summary><i>Previously, jseldess wrote…</i></summary>
> At start of second sentence, "Retryable transaction" should be plural. 
</details>
Done.

---

*[transactions.md, line 128 at r4](https://reviewable.io:443/reviews/cockroachdb/docs/672#-KUs0aYVqGXl5IZ-LV4C:-KUsjW_57xEwUx7tTDBV:b-896fix) ([raw file](https://github.com/cockroachdb/docs/blob/e92f413328c1292ef88d48cbb8bc49c759916264/transactions.md#L128)):*
<details><summary><i>Previously, jseldess wrote…</i></summary>
> In first sentence:
> - change "CockroachDB include" to "CockroachDB include**s**"
> - change "retry transaction" to "retry transaction**s**"
</details>
Done.

---

*[transactions.md, line 218 at r4](https://reviewable.io:443/reviews/cockroachdb/docs/672#-KUs1p_zfqkhAH5n2CJ6:-KUskY1V97WugcUEBV2q:b-896fix) ([raw file](https://github.com/cockroachdb/docs/blob/e92f413328c1292ef88d48cbb8bc49c759916264/transactions.md#L218)):*
<details><summary><i>Previously, jseldess wrote…</i></summary>
> The confusing thing here, to me, is that `REPEATABLE READ` is a synonym for `SNAPSHOT`. I think we nee d to start this section with a clear statement about aliases, as you've done in the `BEGIN` docs, and then go on to explain that, though aliases, they are no equivalent. Thoughts?
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/docs/672)*
<!-- Sent from Reviewable.io -->

cockroach-teamcity commented:
http://cockroach-docs-review.s3-website-us-east-1.amazonaws.com/be3a7c7241be8ea204a0b38e7f3a7ef48de4b952/

sploiselle commented:
@andreimatei Can you make sure that my explanation of the ambiguous errors is legit? [link](http://cockroach-docs-review.s3-website-us-east-1.amazonaws.com/be3a7c7241be8ea204a0b38e7f3a7ef48de4b952/transactions.html#error-handling)

andreimatei commented:
Just a few more comments :)
Sorry for the iterative process, but these are subtle technical points that deserve careful language.

---

Review status: 0 of 17 files reviewed at latest revision, 32 unresolved discussions.

---

*[show-transaction.md, line 8 at r4](https://reviewable.io:443/reviews/cockroachdb/docs/672#-KUrzkyp8yvWZ4qYzh84:-KUtb5Nb61kQ9L56FJFQ:b-ojwblo) ([raw file](https://github.com/cockroachdb/docs/blob/e92f413328c1292ef88d48cbb8bc49c759916264/show-transaction.md#L8)):*
<details><summary><i>Previously, sploiselle (Sean Loiselle) wrote…</i></summary>
> Done.
</details>
As I was talking to Jesse today - technically what happens when you run `SHOW TRANSACTION` outside of of transaction is that you see the priority of the *implicit transaction that was automatically created* for this single statement. This is not quite the "default setting". I'll leave it out to you to judge if this is a distinction without a difference; I, for one, would not introduce any language here about the "default setting", particularly since currently (at least) users can't change the default. I'd rather say that the statement is meaningless if run outside of a transaction.

---

*[transactions.md, line 52 at r5](https://reviewable.io:443/reviews/cockroachdb/docs/672#-KUthTCeWtnsZBtyF8NP:-KUthTCeWtnsZBtyF8NQ:b-o9dxm9) ([raw file](https://github.com/cockroachdb/docs/blob/be3a7c7241be8ea204a0b38e7f3a7ef48de4b952/transactions.md#L52)):*
> ```Markdown
> ## Error Handling
> 
> To handle errors in transactions, you should listen for the following types of errors:
> ```

I think we should clarify that this is only about server-side errors. The different client drivers might throw their own errors in various situations. Notably, "connection lost" errors are also ambiguous when they happen after having sent a COMMIT statement.

---

*[transactions.md, line 56 at r5](https://reviewable.io:443/reviews/cockroachdb/docs/672#-KUsuN6iyQdVLXjZXjcc:-KUsuN6iyQdVLXjZXjcd:bn3lj50) ([raw file](https://github.com/cockroachdb/docs/blob/be3a7c7241be8ea204a0b38e7f3a7ef48de4b952/transactions.md#L56)):*
> ```Markdown
> - **Retryable Errors**: Errors with the code `40001` or string `retry transaction`, which indicate the transaction failed because another concurrent or recent transaction accessed the same values. To handle these errors, you should [retry the transaction](#client-side-transaction-retries).
> 
> - **Ambiguous Errors**: Errors with the code `XX000`, which indicate that the state of the transaction is ambiguous, i.e., you cannot assume it either committed or failed. How you handle these errors depends on how you want to resolve the ambiguity.
> ```

I think it's important to make clear that this ambiguity is limited to errors you get when running a `RELEASE` (or `COMMIT`, when there's no `RELEASE`) statement (or a statement outside of a transaction).
Otherwise, we might strike fear in innocent hearts thinking that all of sudden everything is ambiguous.

This distinction is practically important - the code committing transactions is frequently quite separate from the code that runs inside of a transaction, and deciding what errors need to be handled where is quite important.

---

*[transactions.md, line 60 at r5](https://reviewable.io:443/reviews/cockroachdb/docs/672#-KUtgtyeI9H80IpupnXi:-KUtgtyeI9H80IpupnXj:bbrc843) ([raw file](https://github.com/cockroachdb/docs/blob/be3a7c7241be8ea204a0b38e7f3a7ef48de4b952/transactions.md#L60)):*
> ```Markdown
>   For example, you might want to read values from the database to see if the transaction successfully wrote values before attempting to write the values again or, alternatively, you might write the data again without seeing if the first write attempt succeeded.
> 
> - **SQL Errors**: All other errors, which indicate that a statement in the transaction failed. For example, violating the `UNIQUE` constraint generates an error containing the string `duplicate key value`. After encountering these errors, you can either issue a `COMMIT` or `ROLLBACK` to abort the transaction and revert the database to its state before the transaction began.
> ```

Let's reference the error codes rather than the strings. The problem with the strings is that they can get modified or stripped away completely by client drivers. For example, in Go, I think that for a bunch of common codes (e.g. unique constraint violation), the driver replaces them with its own set of error messages.

---

*[transactions.md, line 105 at r5](https://reviewable.io:443/reviews/cockroachdb/docs/672#-KUtcyHmEfgDFjm7CqeB:-KUtcyHmEfgDFjm7CqeC:b-zb9qvg) ([raw file](https://github.com/cockroachdb/docs/blob/be3a7c7241be8ea204a0b38e7f3a7ef48de4b952/transactions.md#L105)):*
> ```Markdown
>   ~~~
> 
> In these cases, CockroachDB assumes there is nothing conditional about these values, so it can continue to retry the transaction with the values it originally received.
> ```

s/assumes/infers. Let's not make it sound like this assumption might be wrong :)

---

*[transactions.md, line 107 at r5](https://reviewable.io:443/reviews/cockroachdb/docs/672#-KUtdNkQyf4AWjDY8bzZ:-KUtdNkQyf4AWjDY8bz_:b-9e82y8) ([raw file](https://github.com/cockroachdb/docs/blob/be3a7c7241be8ea204a0b38e7f3a7ef48de4b952/transactions.md#L107)):*
> ```Markdown
> In these cases, CockroachDB assumes there is nothing conditional about these values, so it can continue to retry the transaction with the values it originally received.
> 
> If any of the statement's values are conditional (i.e., they change depending on how long the transaction takes to succeed), we recommend instead writing your transaction to use [client-side intervention](#client-side-intervention).
> ```

I don't quite understand what this phrase about conditional values is trying to tell me. I suspect we'd be better advised to not say anything at all.
If the statements inside such a batch depend on timing, the user has already lost even without retries - we make no guarantees about any timing. For example, it could in theory take 5 minutes for just the `BEGIN` stmt to run. Conceptually, when you send a batch like this, the fact that it was retried or not is completely transparent to the client.
If we ever get transaction deadlines (which, coincidentally, in fact I'm just about to suggest again), maybe there's gonna be something to say about the interaction between retries and deadlines that might be related to the point you're trying to make here...

I think we should just say that the transaction might be repeated literally and that might affect the running time (the time it takes a client to receive the results).

---

*[transactions.md, line 143 at r5](https://reviewable.io:443/reviews/cockroachdb/docs/672#-KUtfwSJ2hxeG-0cDeva:-KUtfwSJ2hxeG-0cDevb:b-qjnoni) ([raw file](https://github.com/cockroachdb/docs/blob/be3a7c7241be8ea204a0b38e7f3a7ef48de4b952/transactions.md#L143)):*
> ```Markdown
> - [`ROLLBACK TO cockroach_restart`](rollback-transaction.html#retry-a-transaction) is used when your application detects `40001` / `retry transaction` errors. It provides you a chance to "retry" the transaction by rolling the database's state back to the beginning of the transaction and increasing the transaction's priority. 
>   
>   After issuing `ROLLBACK TO cockroach_restart`, you must issue any statements you want the transaction to contain. Typically, this means recalculating and reissuing a similar set of statements to the previous attempt.
> ```

s/recalculating/recalculating values

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/docs/672#-:-KUtgbITbEnm5a06geEv:b-t5ilzr)*
<!-- Sent from Reviewable.io -->

jseldess commented:
@sploiselle, FYI, `LOCAL` was marked as unimplemented in https://github.com/cockroachdb/cockroach. I removed `LOCAL` from the full grammar and from the `SET TRANSACTION` in https://github.com/cockroachdb/docs/pull/803.

sploiselle commented:
@jseldess Got all of the feedback accounted for

---

Review status: 0 of 17 files reviewed at latest revision, 42 unresolved discussions, some commit checks failed.

---

*[show-transaction.md, line 8 at r4](https://reviewable.io:443/reviews/cockroachdb/docs/672#-KUrzkyp8yvWZ4qYzh84:-KVR4Dlj1yMPEvITlwwE:b-896fix) ([raw file](https://github.com/cockroachdb/docs/blob/e92f413328c1292ef88d48cbb8bc49c759916264/show-transaction.md#L8)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary>
> As I was talking to Jesse today - technically what happens when you run `SHOW TRANSACTION` outside of of transaction is that you see the priority of the *implicit transaction that was automatically created* for this single statement. This is not quite the "default setting". I'll leave it out to you to judge if this is a distinction without a difference; I, for one, would not introduce any language here about the "default setting", particularly since currently (at least) users can't change the default. I'd rather say that the statement is meaningless if run outside of a transaction.
</details>
Done.

---

*[transactions.md, line 52 at r5](https://reviewable.io:443/reviews/cockroachdb/docs/672#-KUthTCeWtnsZBtyF8NP:-KVQMPAISUydwHrnPKS6:b-896fix) ([raw file](https://github.com/cockroachdb/docs/blob/be3a7c7241be8ea204a0b38e7f3a7ef48de4b952/transactions.md#L52)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary>
> I think we should clarify that this is only about server-side errors. The different client drivers might throw their own errors in various situations. Notably, "connection lost" errors are also ambiguous when they happen after having sent a COMMIT statement.
</details>
Done.

---

*[transactions.md, line 56 at r5](https://reviewable.io:443/reviews/cockroachdb/docs/672#-KUsuN6iyQdVLXjZXjcc:-KVQMUNTY8gIUrkr-1Bw:b-896fix) ([raw file](https://github.com/cockroachdb/docs/blob/be3a7c7241be8ea204a0b38e7f3a7ef48de4b952/transactions.md#L56)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary>
> I think it's important to make clear that this ambiguity is limited to errors you get when running a `RELEASE` (or `COMMIT`, when there's no `RELEASE`) statement (or a statement outside of a transaction).
> Otherwise, we might strike fear in innocent hearts thinking that all of sudden everything is ambiguous.
> 
> This distinction is practically important - the code committing transactions is frequently quite separate from the code that runs inside of a transaction, and deciding what errors need to be handled where is quite important.
</details>
Done.

---

*[transactions.md, line 60 at r5](https://reviewable.io:443/reviews/cockroachdb/docs/672#-KUtgtyeI9H80IpupnXi:-KVQOfbDLk2mgx_DCGkc:b-896fix) ([raw file](https://github.com/cockroachdb/docs/blob/be3a7c7241be8ea204a0b38e7f3a7ef48de4b952/transactions.md#L60)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary>
> Let's reference the error codes rather than the strings. The problem with the strings is that they can get modified or stripped away completely by client drivers. For example, in Go, I think that for a bunch of common codes (e.g. unique constraint violation), the driver replaces them with its own set of error messages.
</details>
Done.

---

*[transactions.md, line 105 at r5](https://reviewable.io:443/reviews/cockroachdb/docs/672#-KUtcyHmEfgDFjm7CqeB:-KVQOnQ7NZZoSFhRODrj:b-896fix) ([raw file](https://github.com/cockroachdb/docs/blob/be3a7c7241be8ea204a0b38e7f3a7ef48de4b952/transactions.md#L105)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary>
> s/assumes/infers. Let's not make it sound like this assumption might be wrong :)
</details>
Done.

---

*[transactions.md, line 107 at r5](https://reviewable.io:443/reviews/cockroachdb/docs/672#-KUtdNkQyf4AWjDY8bzZ:-KVRAeu7OaQAKnazv4as:b-896fix) ([raw file](https://github.com/cockroachdb/docs/blob/be3a7c7241be8ea204a0b38e7f3a7ef48de4b952/transactions.md#L107)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary>
> I don't quite understand what this phrase about conditional values is trying to tell me. I suspect we'd be better advised to not say anything at all.
> If the statements inside such a batch depend on timing, the user has already lost even without retries - we make no guarantees about any timing. For example, it could in theory take 5 minutes for just the `BEGIN` stmt to run. Conceptually, when you send a batch like this, the fact that it was retried or not is completely transparent to the client.
> If we ever get transaction deadlines (which, coincidentally, in fact I'm just about to suggest again), maybe there's gonna be something to say about the interaction between retries and deadlines that might be related to the point you're trying to make here...
> 
> I think we should just say that the transaction might be repeated literally and that might affect the running time (the time it takes a client to receive the results).
</details>
Done.

---

*[transactions.md, line 143 at r5](https://reviewable.io:443/reviews/cockroachdb/docs/672#-KUtfwSJ2hxeG-0cDeva:-KVR49tx7HPXkdjaf44-:b-896fix) ([raw file](https://github.com/cockroachdb/docs/blob/be3a7c7241be8ea204a0b38e7f3a7ef48de4b952/transactions.md#L143)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary>
> s/recalculating/recalculating values
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/docs/672#-:-KUsndQavRU2bXEOS8qt:bt7r8ff)*
<!-- Sent from Reviewable.io -->

cockroach-teamcity commented:
http://cockroach-docs-review.s3-website-us-east-1.amazonaws.com/63fd979bdfe929dee7260da4e7da2083adab2dc7/

cockroach-teamcity commented:
http://cockroach-docs-review.s3-website-us-east-1.amazonaws.com/4e177c194b057a8dea10826290935a5d1c94a059/

cockroach-teamcity commented:
http://cockroach-docs-review.s3-website-us-east-1.amazonaws.com/b9ac706a7969c742b9ef3dc71b98dd77e9fda6cf/

sploiselle commented:
@jseldess Two changes in the [Automatic Retries](http://cockroach-docs-review.s3-website-us-east-1.amazonaws.com/b9ac706a7969c742b9ef3dc71b98dd77e9fda6cf/transactions.html#automatic-retries) section:

- Line beginning "However, if the transaction..."
- Line beginning "It’s also important to note..."

If both of those look good, think we can merge.

---

Review status: 0 of 17 files reviewed at latest revision, 41 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/docs/672#-:-KVRLxiawnEoIOSZCG8Y:b-523yig)*
<!-- Sent from Reviewable.io -->

jseldess commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 5 of 16 files at r1, 3 of 12 files at r2, 1 of 6 files at r4, 4 of 6 files at r5, 4 of 4 files at r6.
Review status: all files reviewed at latest revision, 41 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/docs/672#-:-KVRS24Oz2WeDPAAZLw8:bnfp4nl)*
<!-- Sent from Reviewable.io -->
